pkg/core/workflow/types.go

```diff
@@ -0,0 +1,164 @@
+// Copyright 2021 Chaos Mesh Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package workflow includes all type definitaions about Workflow.
+package workflow
+
+import (
+	"github.com/jinzhu/gorm"
+)
+
+// Workflow defines the root structure of a workflow.
+type Workflow struct {
+	gorm.Model
+	UID         string         `gorm:"index:uid" json:"uid"`
+	Name        string         `json:"name"`
+	Entry       string         `json:"entry"` // the entry node name
+	Status      workflowStatus `json:"status"`
+	CurrentNode TopologyNode   `json:"current_node"`
+	Topology    Topology       `json:"topology"`
+}
+
+// workflowState defines the current state of a workflow.
+//
+// Includes: Initializing, Running, Errored, Finished.
+//
+// Const definitions can be found below this type.
+type workflowStatus string
+
+const (
+	// WorkflowInitializing represents a workflow is being initialized.
+	WorkflowInitializing workflowStatus = "Initializing"
+
+	// WorkflowRunning represents that a workflow is running.
+	WorkflowRunning workflowStatus = "Running"
+
+	// WorkflowErrored represents an error in a workflow.
+	WorkflowErrored workflowStatus = "Errored"
+
+	// WorkflowFinished represents that a workflow has ended.
+	WorkflowFinished workflowStatus = "Finished"
+)
+
+// Topology describes the process of a workflow.
+type Topology struct {
+	Nodes []TopologyNode `json:"nodes"`
+}
+
+// TopologyNode defines the basic structure of a node.
+type TopologyNode struct {
+	Type  nodeType  `json:"type"`
+	State nodeState `json:"state,omitempty"`
+}
+
+// Node defines the single step of a workflow.
+type Node struct {
+	TopologyNode
+	Serial   NodeSerial   `json:"serial,omitempty"`
+	Parallel NodeParallel `json:"parallel,omitempty"`
+	Template Template     `json:"template"`
+}
+
+// NodeSerial defines SerialNode's specific fields.
+type NodeSerial struct {
+	Tasks    []Node `json:"tasks"`
+	Circular bool   `json:"circular"`
```

Comment:
What does the `Circular` field mean?

 Is this field affect the drawing of workflow?

Reply:
> Is this field affect the drawing of workflow?

Not affect this.

I remembered that we are talking about the serial node that can be re-executed when it's ended. So I add `Circular ` to this struct.

But I now realize this kind of loop execution can be controlled by the task node. I don't know if I understand it right. If it is, this field is no needed.

Reply:
OK. 🙈 

As you say, it's a little complex to determine that field is true or false.

I think it's safe to delete it.

Reply:
👌🏽